ENH: add `signal.remezord`

luxedo opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes PR https://github.com/scipy/scipy/pull/4122

#### What does this implement/fix?
<!--Please explain your changes.-->
Picks up previous `remezord` implementation from @nils-werner and adds the required documentation for publishing.

#### Additional information
<!--Any additional information you think is important.-->
Tasks to complete this PR:

- [x] Lint: Remove lambdas
- [x] code needs unit tests.
- [x] all functions except remezord need to start with an underscore.
- [x] the docstrings for remlplen_* repeat the same text 3 times. The 2nd and 3rd docstring can refer to the first one. The references to papers should all be moved to the remezord docstring.
- [x] the code needs some PEP8 cleanup (spaces around operators etc.)
- [x] the Hz parameter is missing from the docstring.
- [x] docstring format: See should be See Also, typo in Paramters, remove blank lines right below section headings.
- [x] An example in the docstring would be very useful.
- [x] remezord should be added to the function listing in signal/__init__.py.

nils-werner committed:
remezord implementation

luxedo committed:
ENH: move remezord to appropriate module

luxedo committed:
DOC: add remezord fs argument description

WarrenWeckesser commented:
Thanks @luxedo.  I don't think I'll have time for a full review in the short term, but here's one item that must be fixed: change the parameter name `Hz` to `fs`.  We adopted this convention throughout the `signal` module several years ago (see, for example this [section of the 1.2.0 release notes](https://docs.scipy.org/doc/scipy/release/1.2.0-notes.html#scipy-signal-improvements)).

luxedo commented:
@WarrenWeckesser that's exactly what I just did when adding the docstring for the ~`Hz`~ `fs` argument
![image](https://github.com/scipy/scipy/assets/14118472/60def5e7-d57d-4c7c-999b-c6c9bac53280)

luxedo committed:
DOC: update remezord docstrings

luxedo committed:
BUG: remove lambdas from remezord

luxedo committed:
DOC: add remezord to signal.__init__

luxedo committed:
DOC: typo in remezord

luxedo committed:
DOC: move remezord references out of inner functions

luxedo committed:
TST: add remezord unittests

luxedo committed:
DOC: add remezord examples

luxedo committed:
DOC: fix remezord docstrings underline

luxedo committed:
DOC: fix remezord docstrings backticks

luxedo committed:
DOC: fix remezord examples missing imports

luxedo commented:
I have a version that's ready for review. I'm not a `signals` guy but I tried to make things consistent, so a little feedback here is more than welcome.

larsoner commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,248 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+    >>> desired
+    [1., 0.]
+    >>> weights
+    [1., 17.09973573]
```
It would be nice to create the resulting filter and plot its frequency response using `freqz`... then you could at least eyeball that the results are consistent. Bonus points if you can make it look like Example 1 from the [linked website](http://www.ece.northwestern.edu/local-apps/matlabhelp/toolbox/signal/remezord.html)!

luxedo replied:
I added the plot and it seems ok for me:

https://github.com/scipy/scipy/pull/20983/files#diff-c65308cc23ecff9cc1a62529e3298c1f26440a5fa95787fee881449ef8f1843bR1037-R1051

I also created a second example based on `remez` _pass-band_ filter example.

larsoner commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -481,6 +481,70 @@ def test_fs_validation(self):
+class TestRemezord:
+    """
+    Test examples taken from:
+    http://www.ece.northwestern.edu/local-apps/matlabhelp/toolbox/signal/remezord.html
```
I don't see these numbers on that page, so assuming the reference values were computed in MATLAB or Octave can you add that to the comment/doc here? Like:

```suggestion
    http://www.ece.northwestern.edu/local-apps/matlabhelp/toolbox/signal/remezord.html
    with reference values computed in MATLAB.
```
At least in MATLAB I get:
```MATLAB
>> format compact
>> format long
>> rp = 3;          % Passband ripple
rs = 40;         % Stopband ripple
fs = 2000;       % Sampling frequency
f = [500 600];   % Cutoff frequencies
a = [1 0];       % Desired amplitudes
% Compute deviations
dev = [(10^(rp/20)-1)/(10^(rp/20)+1)  10^(-rs/20)]; 
[n,fo,ao,w] = remezord(f,a,dev,fs);
>> n, fo, ao, w
n =
    22
fo =
                   0
   0.500000000000000
   0.600000000000000
   1.000000000000000
ao =
     1
     1
     0
     0
w =
   1.000000000000000
  17.0997357343619040
```

Also the Example 2 isn't in there, but locally I ran it in MATLAB and get the following, can it also become a test case?
```MATLAB
>> [n,fo,ao,w] = remezord([1500 2000],[1 0],[0.01 0.1],8000 );
>> n, fo, ao, w
n =
    21
fo =
         0
    0.3750
    0.5000
    1.0000
ao =
     1
     1
     0
     0
w =
    10
     1
```

luxedo replied:
Added Example 2:

https://github.com/scipy/scipy/pull/20983/files#diff-bb7c811e81932922f4a8964acc03ac3633a245fbc07a6120d4c814d51d2b810fR550-R559

larsoner reviewed (commented):
I don't know the implementation specifics, but the code at least looks clean and easy enough to follow. One suggestion for improving the example and I have one more test case that can hopefully easily be added!

luxedo committed:
TST: Add remezord test reference quote

Co-authored-by: Eric Larson <larson.eric.d@gmail.com>

luxedo committed:
DOC: Add filter creation to `remezord` examples

Also add a second band-pass example

luxedo committed:
TST: add remezord example 2

j-bowhay commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -481,6 +481,84 @@ def test_fs_validation(self):
+        amps = np.array([40, 50, 60])
+        rips = np.array([3, 4, 5])
+        # Nonexistent algorithm
+        assert_raises(ValueError, remezord, freqs, amps, rips, alg="no_alg")
```
For new tests please could we use the context manager version
```suggestion
        with assert_raises(ValueError):
            remezord(freqs, amps, rips, alg="no_alg")
```

j-bowhay commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -481,6 +481,84 @@ def test_fs_validation(self):
+        assert numtaps == 27
+        assert_equal(bands, [0, 0.25, 0.3, 0.5])
+        assert_equal(desired, [1., 0.])
+        assert_almost_equal(weights, [1., 17.09973573])
```
Here and elsewhere please could you use `assert_allclose` instead
```suggestion
        assert_allclose(weights, [1., 17.09973573])
```

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,296 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+    return int(N4)
+
+
+def remezord(freqs, amps, rips, fs=1.0, alg="ichige"):
```
We should consider making some of these keyword only as per https://docs.scipy.org/doc/scipy/dev/missing-bits.html#required-keyword-names

Please feel free to suggest a better location of the `*`
```suggestion
def remezord(freqs, amps, rips, *, fs=1.0, alg="ichige"):
```

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,296 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+
+    >>> def plot_response(w, h, title):
+    ...     "Utility function to plot response functions"
+    ...     fig = plt.figure()
+    ...     ax = fig.add_subplot(111)
```
```suggestion
    ...     fig, ax = plt.subplots()
```

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,296 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+    Raises
+    ------
+    ValueError:
+        - If any element in ``freqs`` is negative, greater than 0.5, or the
```
Parameters should be in single backticks
```suggestion
        - If any element in `freqs` is negative, greater than 0.5, or the
```

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,296 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+    ValueError:
+        - If any element in ``freqs`` is negative, greater than 0.5, or the
+          length is not valid.
+        - If the lengths of ``amps`` and ``rips`` are not equal or one more
+          than half the length of ``freqs``.
+        - If any element in ``rips`` is negative.
+        - If ``alg`` is not a valid string option.
```
```suggestion
        - If the lengths of `amps` and `rips` are not equal or one more
          than half the length of `freqs`.
        - If any element in `rips` is negative.
        - If `alg` is not a valid string option.
```

j-bowhay commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,296 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+    # ripple and all of the other ripples:
+    weight = max(rips) / rips
+
+    return [L, bands, amps, weight]
```
Is there a reason for returning a list over the more conventional tuple?
```suggestion
    return L, bands, amps, weight
```

j-bowhay commented on scipy/signal/tests/test_fir_filter_design.py:
```diff
@@ -481,6 +481,84 @@ def test_fs_validation(self):
+        assert_equal(desired, [1., 0.])
+        assert_almost_equal(weights, [1., 17.09973573])
+
+    def test_remezord_example1_herrmann(self):
+        rp, rs = 3, 40
+        fs = 2000
+        freqs = np.array([500, 600])
+        amps = np.array([1, 0])
+        rips = [(10**(rp/20)-1)/(10**(rp/20)+1), 10**(-rs/20)]
+        numtaps, bands, desired, weights = remezord(freqs, amps, rips, fs=fs,
+                                                    alg="herrmann")
+        assert numtaps == 25
+        assert_equal(bands, [0, 0.25, 0.3, 0.5])
+        assert_equal(desired, [1., 0.])
+        assert_almost_equal(weights, [1., 17.09973573])
+
+    def test_remezord_example1_kaiser(self):
+        rp, rs = 3, 40
+        fs = 2000
+        freqs = np.array([500, 600])
+        amps = np.array([1, 0])
+        rips = [(10**(rp/20)-1)/(10**(rp/20)+1), 10**(-rs/20)]
+        numtaps, bands, desired, weights = remezord(freqs, amps, rips, fs=fs,
+                                                    alg="kaiser")
+        assert numtaps == 23
+        assert_equal(bands, [0, 0.25, 0.3, 0.5])
+        assert_equal(desired, [1., 0.])
+        assert_almost_equal(weights, [1., 17.09973573])
+
+    def test_remezord_example2(self):
+        fs = 8000
+        freqs = np.array([1500, 2000])
+        amps = np.array([1, 0])
+        rips = [0.01, 0.1]
+        numtaps, bands, desired, weights = remezord(freqs, amps, rips, fs=fs)
+        assert numtaps == 24
+        assert_equal(bands, [0, 0.1875, 0.25, 0.5])
+        assert_equal(desired, [1., 0.])
+        assert_almost_equal(weights, [10, 1])
```
Optional but since these tests all follow the same structure it would be nice to have instead one function that we parameterize

j-bowhay reviewed (changes requested):
Thanks @luxedo some stylistic comments

luxedo committed:
TST: parameterize remezord example 1

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

luxedo committed:
DOC: remezord docstring stylistic changes

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

luxedo commented:
I believe I have fixed all the issues:
- [x] [Add context manager `assert_raises`](https://github.com/scipy/scipy/pull/20983/files#diff-bb7c811e81932922f4a8964acc03ac3633a245fbc07a6120d4c814d51d2b810fR499-R515)
- [x] [Update assertions to `assert_allclose`](https://github.com/scipy/scipy/pull/20983/files#diff-bb7c811e81932922f4a8964acc03ac3633a245fbc07a6120d4c814d51d2b810fR530-R533)
- [x] [Mark parameters after `rip` as keyword only](https://github.com/scipy/scipy/pull/20983/files#diff-c65308cc23ecff9cc1a62529e3298c1f26440a5fa95787fee881449ef8f1843bR934)
- [x] [Simplify fig/ax creation in example plot](https://github.com/scipy/scipy/pull/20983/files#diff-c65308cc23ecff9cc1a62529e3298c1f26440a5fa95787fee881449ef8f1843bR1043)
- [x] [Remove double backticks](https://github.com/scipy/scipy/pull/20983/files#diff-c65308cc23ecff9cc1a62529e3298c1f26440a5fa95787fee881449ef8f1843bR942-R969)
- [x] [Return tuple instead of list](https://github.com/scipy/scipy/pull/20983/files#diff-c65308cc23ecff9cc1a62529e3298c1f26440a5fa95787fee881449ef8f1843bR1128)
- [x] [Parameterize example 1 tests ](https://github.com/scipy/scipy/pull/20983/files#diff-bb7c811e81932922f4a8964acc03ac3633a245fbc07a6120d4c814d51d2b810fR517-R533)

luxedo commented:
Hello! Any updates on this?

DietBru commented on scipy/signal/_fir_filter_design.py:
```diff
@@ -834,6 +835,299 @@ def remez(numtaps, bands, desired, *, weight=None, type='bandpass',
+def _oddceil(x):
+    """Return the smallest odd integer not less than x."""
+
+    return _oddround(x + 1)
```
A matter of personal taste: The functions `_oddround` and `_oddceil` are each only used one time in the following code. In-lining these functions (inlcuding a comment) would make it a tiny bit easier to reason about the code.

DietBru reviewed (changes requested):
Sorry for not finding the time to look at your PR earlier. The code looks plausible and is easy to follow (I am no expert on the Remez algorithm).

What is missing, in my opinion, is some guidance for the user when to chose which algorithm. Ichige's paper might give some pointers here.

The other comments are just small nitpicks on coding style, which are more of personal taste things.

luxedo committed:
DOC: remezord docstring stylistic changes

Co-authored-by: Dietrich Brunn <Dietrich.Brunn@web.de>

luxedo committed:
DOC: remezord detailed notes for each method

luxedo committed:
DOC: update remezord test assertions to xp_assert_close

luxedo committed:
TST: fix desired dtype

luxedo commented:
@DietBru I addressed your suggestions. For the Notes about usage I don't feel I'm qualified to add 

https://github.com/scipy/scipy/pull/20983/files#diff-c65308cc23ecff9cc1a62529e3298c1f26440a5fa95787fee881449ef8f1843bR981-R999

DietBru commented:
Sorry for not finding the time to spend time on this PR. 
I did a little bit of googling and did not find to many resources. E.g., Matlab seems only to implement Hermann's method in [firpmord]. There are some more recent [articles] citing Ischige's article on this topic. 

Notably, I did not stumble across anything recommending Hermann's or Kaiser's method over Ichige's algorithm (as also seen in the Notes). Due to the age of the references, I would vote to include only Ichige's method and exclude the other two. If reasons come up to add them later, it still can be done in a separate PR (removing an option is always a lot more involving. 

If you want @luxedo, I can make those changes to this PR—but I would be also more than perfectly fine, if you would update it.

[firpmord]: https://de.mathworks.com/help/signal/ref/firpmord.html
[articles]: https://scholar.google.com/scholar?cites=5304257917270837297

luxedo commented:
I can take a look by the end of this month

DietBru committed:
Signal.remezord: Improve docstr and rename parameters.

* Rename parameters to be more descriptive. Another reason for renaming was that in the HTML docs `freqs` links to `signal.freqs`, which cannot be easily suppressed.

* Improved notes section:
  - Explain difference to `remez` function
  - Emphasize that only an estimate is provided

* Beautify example section a bit.

* Fixed `numtaps` check in `TestRemezord.test_remezord_example1`.

DietBru committed:
Small fix in `signal.remez` which fixes an exception caused by an example in the `signal.remezord` docstr.
